File: engine/src/math/matrix.cpp
#include "engine/math/matrix.hpp"

#include <glm/glm.hpp>

namespace mana {
    glm::mat4 toGLM(const Matrix<float, 4, 4> &mat) {
        glm::mat4 ret;
        for (int x = 0; x < 4; x++) {
            for (int y = 0; y < 4; y++) {
                ret[x][y] = mat.get(x, y);
            }
        }
        return ret;
    }

    Matrix<float, 4, 4> toMana(const glm::mat4 &mat) {
        Matrix<float, 4, 4> ret;
        for (int x = 0; x < 4; x++) {
            for (int y = 0; y < 4; y++) {
                ret.set(x, y, mat[x][y]);
            }
        }
        return ret;
    }

    glm::vec4 toGLM(const Vec4f &vec) {
        return glm::vec4(vec.x, vec.y, vec.z, vec.w);
    }

    Vec4f toMana(const glm::vec4 &vec) {
        return Vec4f(vec.x, vec.y, vec.z, vec.w);
    }

    Vector4<float> operator*(const Matrix<float, 4, 4> &lhs, const Vector4<float> &rhs) {
        /*glm::mat4::col_type ret = toGLM(rhs) * toGLM(lhs);
        return toMana(ret);*/
        Vector4<float> ret;
        for (int row = 0; row < lhs.height(); row++) {
            float *out = nullptr;
            switch (row) {
                case 0:
                    out = &ret.x;
                    break;
                case 1:
                    out = &ret.y;
                    break;
                case 2:
                    out = &ret.z;
                    break;
                case 3:
                    out = &ret.w;
                    break;
                default:
                    assert(false);
            }

            assert(out != nullptr);

            for (int column = 0; column < lhs.width(); column++) {
                float in;
                switch (column) {
                    case 0:
                        in = rhs.x;
                        break;
                    case 1:
                        in = rhs.y;
                        break;
                    case 2:
                        in = rhs.z;
                        break;
                    case 3:
                        in = rhs.w;
                        break;
                    default:
                        assert(false);
                }

                *out += lhs.get(column, row) * in;
            }
        }
        return ret;
    }

    Matrix<float, 4, 4> operator*(const Matrix<float, 4, 4> &lhs, const Matrix<float, 4, 4> &rhs) {
        glm::mat4 lh = toGLM(lhs);
        glm::mat4 rh = toGLM(rhs);
        return toMana(lh * rh);
    }
}

File: engine/src/csharp/mana/math/matrix.cs
/**
 *  Mana - 3D Game Engine
 *  Copyright (C) 2021  Julian Zampiccoli
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

namespace Mana
{
    /**
    A 4 by 4 matrix of floats, stored row major.
    **/
    public class Matrix
    {
        public float this[int column, int row]
        {
            get => data[row * 4 + column];
            set => data[row * 4 + column] = value;
        }

        private readonly float[] data = new float[16];

        public void Set(int column, int row, float value)
        {
            this[column, row] = value;
        }

        public float Get(int column, int row)
        {
            return this[column, row];
        }

        public static Vector4 operator *(Matrix mat, Vector4 vec)
        {
            Vector4 ret = new Vector4();
            for (int row = 0; row < 4; row++)
            {
                float r = 0;
                switch (row)
                {
                    case 0:
                        r = ret.x;
                        break;
                    case 1:
                        r = ret.y;
                        break;
                    case 2:
                        r = ret.z;
                        break;
                    case 3:
                        r = ret.w;
                        break;
                    default:
                        break;
                }
                for (int column = 0; column < 4; column++)
                {
                    float f = 0;
                    switch (column)
                    {
                        case 0:
                            f = vec.x;
                            break;
                        case 1:
                            f = vec.y;
                            break;
                        case 2:
                            f = vec.z;
                            break;
                        case 3:
                            f = vec.w;
                            break;
                        default:
                            break;
                    }
                    r += mat[column, row] * f;
                }
                switch (row)
                {
                    case 0:
                        ret.x = r;
                        break;
                    case 1:
                        ret.y = r;
                        break;
                    case 2:
                        ret.z = r;
                        break;
                    case 3:
                        ret.w = r;
                        break;
                    default:
                        break;
                }
            }
            return ret;
        }
    }
}